### PR TITLE
[TRTLLM-12950][perf] DSv4 follow-up: DeepGEMM and MegaMoE

### PR DESCRIPTION
**Pull request by lfr-0531**
## Summary by CodeRabbit

* **New Features**
  * Added a new MoE preparation path that can directly set up selected experts, weights, and quantized activations.
  * Added support for bounded SwiGLU behavior in several fused MoE kernels and Python APIs.
  * Added a helper to detect when all-reduce inputs already use an NCCL window buffer.

* **Bug Fixes**
  * Improved MoE scheduling and backend selection for fused-prepare and routing-separation cases.
  * Reduced unnecessary memory retention by releasing CPU-backed weights more aggressively.
  * Refined all-reduce caching behavior for NCCL window-backed inputs.

## Description

This PR carves the PR-5 DeepGEMM/MegaMoE follow-up scope from `github/feat/deepseek_v4` after `a7728aa` onto current `github/main`.

It includes:
- DeepGEMM dependency and PDL setup.
- MegaMoE prepare/custom-op registration and scheduler follow-up.
- MegaMoE loader/OOM and EPLB-related follow-up that can be applied without introducing the full DSv4 model file.
- CuTeDSL grouped-GEMM swiglu-limit follow-up and related unit coverage.

It intentionally excludes NVFP4 MegaMoE backend/tuning commits `aad9e0f008` and `f90246b3d8` because equivalent material is already on `main`, and it excludes the `modeling_deepseekv4.py` hunks because that file is not present on current `main` and belongs to the PR-3 DSv4 model/sparse-attention stack.

Rebase/conflict notes are recorded in `tmp/dsv4_final_prs/PR5_rebase_conflicts_20260625.md`.

## Test Coverage

- Build/install:
  - `python3 ./scripts/build_wheel.py --trt_root /usr/local/tensorrt --benchmarks --use_ccache --cuda_architectures "90-real;100-real" --configure_cmake`
  - `.venv-3.12/bin/pip install --force-reinstall --no-deps build/tensorrt_llm-1.3.0rc20-cp312-cp312-linux_x86_64.whl`
  - Verified installed `libth_common.so` contains the PR-5 C++ tactic replacement string.
- Static/scope:
  - `git diff --check`
  - Strict changed-file conflict-marker search.
  - PR-5 forbidden-scope check returned no matches.
  - `pre-commit run --files ...` passed on touched source/test files.
- Unit tests:
  - Focused installed-wheel MoE/CUTLASS/MegaMoE regression:
    - `6 passed, 3 warnings`
  - `PATH="$PWD/.venv-3.12/bin:$PATH" CUDA_VISIBLE_DEVICES=1 timeout 12m .venv-3.12/bin/python -m pytest tests/unittest/_torch/thop/parallel/test_cute_dsl_moe.py tests/unittest/_torch/thop/parallel/test_fp8_quantize.py -q --tb=short`
    - `934 passed, 9 skipped, 245 warnings`

Full `tests/unittest/_torch/modules/moe` caveat:
- One full-suite attempt before the final verification fixes exited with CUDA illegal-memory-access cascading errors after the CUDA context was poisoned.
- A diagnostic rerun with `CUDA_LAUNCH_BLOCKING=1`, `-x`, and a 15 minute timeout did not reproduce a first failure before timeout.

## PR Checklist

Please review the following before submitting your PR:
- PR description clearly explains what and why. If using CodeRabbit's summary, please make sure it makes sense.
- PR Follows [TRT-LLM CODING GUIDELINES](https://github.com/NVIDIA/TensorRT-LLM/blob/main/CODING_GUIDELINES.md) to the best of your knowledge.
- Test cases are provided for new code paths (see [test instructions](https://github.com/NVIDIA/TensorRT-LLM/tree/main/tests#1-how-does-the-ci-work))
- If PR introduces API changes, an appropriate PR label is added - either `api-compatible` or `api-breaking`. For `api-breaking`, include `BREAKING` in the PR title.
- Any new dependencies have been scanned for license and vulnerabilities
- [CODEOWNERS](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/CODEOWNERS) updated if ownership changes
- Documentation updated as needed
- Update [tava architecture diagram](https://github.com/NVIDIA/TensorRT-LLM/blob/main/.github/tava_architecture_diagram.md) if there is a significant design change in PR.
- The reviewers assigned automatically/manually are appropriate for the PR.

- [x] Please check this after reviewing the above items as appropriate for this PR.

## GitHub Bot Help

To see a list of available CI bot commands, please comment `/bot help`.

### CODE REVIEW

File: tensorrt_llm/_torch/modules/fused_moe/create_moe.py [hunk 1]
```diff
@@ -334,13 +334,19 @@ def create_moe_backend(
...
             CutlassFusedMoE, TritonFusedMoE, TRTLLMGenFusedMoE, WideEPMoE,
-            DeepGemmFusedMoE, MegaMoECuteDsl
+            DeepGemmFusedMoE, MegaMoECuteDsl, MegaMoEDeepGemm, CuteDslFusedMoE
         ], f"swiglu_limit is not supported in {moe_cls.__name__}."
+        if moe_cls in (MegaMoEDeepGemm,
+                       CuteDslFusedMoE) and swiglu_limit is not None:
+            raise NotImplementedError(
+                f"{moe_cls.__name__} supports swiglu_limit_scalar (uniform "
+                "clamp) only; per-expert swiglu_limit tensor is not "
+                "supported. Set swiglu_limit=None.")
```
Comment (xxi-nv) [inline tensorrt_llm/_torch/modules/fused_moe/create_moe.py]: From this comment, the MegaMoEDeepGemm and CuteDslFusedMoE should be removed from the assertion list.

File: tests/unittest/_torch/modules/moe/test_moe_scheduler.py [hunk 2]
```diff
@@ -0,0 +1,237 @@
...
+import torch
+
+from tensorrt_llm._torch.modules.fused_moe.fused_moe_trtllm_gen import TRTLLMGenFusedMoE
+from tensorrt_llm._torch.modules.fused_moe.moe_scheduler import (
+    ExternalCommMoEScheduler,
+    FusedCommMoEScheduler,
+)
+
+
+class _SeparatedRouting:
```
Comment (xxi-nv) [inline tests/unittest/_torch/modules/moe/test_moe_scheduler.py]: Do we really need this test file?

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py [hunk 3]
```diff
@@ -4473,6 +4473,17 @@ def process_weights_after_loading(self,
         # Finalize shared expert alphas and fc31_scale_c for online EPLB
         self._finalize_shared_expert_alphas(module)
 
+        # Cubin clamp / GLU bias inputs are consumed in the pre-dequant GEMM
```
Comment (xxi-nv) [inline tensorrt_llm/_torch/modules/fused_moe/quantization.py]: This will also change the logic in W4A8NVFP4FP8TRTLLMGenFusedMoEMethod, which doesn't have this process before, could you help to confirm?

Comment (xxi-nv) [reply tensorrt_llm/_torch/modules/fused_moe/quantization.py]: Maybe it is a potential bug, bug fixed by this PR.

File: tensorrt_llm/_torch/modules/fused_moe/mega_moe/mega_moe_deepgemm.py [hunk 4]
```diff
@@ -79,6 +80,14 @@ def _trtllm_mxfp8_quantize_available() -> bool:
     return hasattr(torch.ops, "trtllm") and hasattr(torch.ops.trtllm, "mxfp8_quantize")
 
 
+def _trtllm_megamoe_prepare_available() -> bool:
```
Comment (xxi-nv) [inline tensorrt_llm/_torch/modules/fused_moe/mega_moe/mega_moe_deepgemm.py]: Could we always run this path?

**Review by xxi-nv — APPROVED** [review]
The MOE LGTM and leave some comments.

File: tensorrt_llm/_torch/modules/fused_moe/create_moe.py (hunk 1 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/modules/fused_moe/create_moe.py]: Fixed! Thanks, Xingfei!

File: tests/unittest/_torch/modules/moe/test_moe_scheduler.py (hunk 2 above)
Comment (lfr-0531) [reply tests/unittest/_torch/modules/moe/test_moe_scheduler.py]: Removed this test file.

File: tensorrt_llm/_torch/modules/fused_moe/quantization.py (hunk 3 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/modules/fused_moe/quantization.py]: I'm not 100% sure about this. Just to be safe, I've moved this logic over to `NVFP4TRTLLMGenFusedMoEMethod`. @Barry-Delaney, plz help double check.

File: tensorrt_llm/_torch/modules/fused_moe/mega_moe/mega_moe_deepgemm.py (hunk 4 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/modules/fused_moe/mega_moe/mega_moe_deepgemm.py]: Updated.

File: tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py [hunk 5]
```diff
@@ -401,6 +401,18 @@ def _(
         dtype = out_dtype if out_dtype is not None else torch.bfloat16
         return act.new_empty((m, n), dtype=dtype)
 
+    @torch.library.register_fake("trtllm::megamoe_prepare")
```
Comment (liji-nv) [inline tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py]: No need fake impl for inplace op. Better add to inplace_info().

**Review by liji-nv — APPROVED** [review]

**Review by lancelly — APPROVED** [review]
Pyexecutor part LGTM

File: tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py (hunk 5 above)
Comment (lfr-0531) [reply tensorrt_llm/_torch/custom_ops/cpp_custom_ops.py]: Fixed in https://github.com/NVIDIA/TensorRT-LLM/pull/15632/commits/8c994fd6e1b1a6a04df27a961bdc4a025f89c21a. Thanks, Jin!
